astropy/coordinates/representation/base.py

```diff
@@ -377,14 +382,34 @@ def _apply(self, method, *args, **kwargs):
         return new
 
     def __setitem__(self, item, value):
-        if value.__class__ is not self.__class__:
+        set_mask = value is np.ma.masked
+        clear_mask = value is np.ma.nomask
+        if not (value.__class__ is self.__class__ or set_mask or clear_mask):
             raise TypeError(
                 "can only set from object of same class: "
                 f"{self.__class__.__name__} vs. {value.__class__.__name__}"
+                " (unless setting or clearing the mask with"
+                " np.ma.masked or np.ma.nomask)."
             )
 
+        if not self.masked:
+            if clear_mask:
+                # Clearing masked elements on an unmasked instance: nothing to do.
+                return
+
+            # Ensure our components are masked if a mask needs to be set.
+            if set_mask or value.masked:
+                self._ensure_masked()
+
+        if set_mask or clear_mask:
+            for comp in self.components:
```

Comment:
Unrelated to this PR but maybe we should have a `_components` attribute that provides an iterator for the `_`-prefixed components.